homeassistant/components/synology_dsm/backup.py

```diff
@@ -0,0 +1,223 @@
+            del hass.data[DATA_BACKUP_AGENT_LISTENERS]
+
+    return remove_listener
+
+
+class SynologyDSMBackupAgent(BackupAgent):
+    """Synology DSM backup agent."""
+
+    domain = DOMAIN
+
+    def __init__(self, hass: HomeAssistant, entry: ConfigEntry) -> None:
+        """Initialize the Synology DSM backup agent."""
+        super().__init__()
+        LOGGER.debug("Initializing Synology DSM backup agent for %s", entry.unique_id)
+        self.name = entry.title
+        self.path = (
+            f"{entry.options[CONF_BACKUP_SHARE]}/{entry.options[CONF_BACKUP_PATH]}"
+        )
+        syno_data: SynologyDSMData = hass.data[DOMAIN][entry.unique_id]
+        self.api = syno_data.api
+
+    @property
+    def _file_station(self) -> SynoFileStation:
+        if TYPE_CHECKING:
+            # we ensure that file_station exist already in async_get_backup_agents
+            assert self.api.file_station
+        return self.api.file_station
+
+    async def async_download_backup(
+        self,
+        backup_id: str,
+        **kwargs: Any,
+    ) -> AsyncIterator[bytes]:
+        """Download a backup file.
+
+        :param backup_id: The ID of the backup that was returned in async_list_backups.
+        :return: An async iterator that yields bytes.
+        """
+        try:
+            resp = await self._file_station.download_file(
+                path=self.path,
+                filename=f"{backup_id}.tar",
+            )
+        except SynologyDSMAPIErrorException as err:
+            raise BackupAgentError("Failed to download backup") from err
+
+        if TYPE_CHECKING:
+            assert isinstance(resp, StreamReader)
+
+        return ChunkAsyncStreamIterator(resp)
+
+    async def async_upload_backup(
+        self,
+        *,
+        open_stream: Callable[[], Coroutine[Any, Any, AsyncIterator[bytes]]],
+        backup: AgentBackup,
+        **kwargs: Any,
+    ) -> None:
+        """Upload a backup.
+
+        :param open_stream: A function returning an async iterator that yields bytes.
+        :param backup: Metadata about the backup that should be uploaded.
+        """
+
+        # upload backup.tar file first
+        try:
+            await self._file_station.upload_file(
+                path=self.path,
+                filename=f"{backup.backup_id}.tar",
+                source=await open_stream(),
+                create_parents=True,
+            )
+        except SynologyDSMAPIErrorException as err:
+            raise BackupAgentError("Failed to upload backup") from err
+
+        # upload backup_meta.json file when backup.tar was successful uploaded
+        try:
+            await self._file_station.upload_file(
+                path=self.path,
+                filename=f"{backup.backup_id}_meta.json",
+                source=json_dumps(backup.as_dict()).encode(),
+            )
+        except SynologyDSMAPIErrorException as err:
+            raise BackupAgentError("Failed to upload backup") from err
```

Comment:
Are these files visible to the user? If they are, things will go south if the user renames the files, or removes one of them. In that case, it's generally not safe to store metadata and backup content separately, and also not to assume the filenames matches a specific pattern.
OTOH, if these files can't be accessed directly by the user this is OK.

Reply:
as discussed on discord, the file-station api does not provide any the possibility to additional meta data to uploaded files - at the end we are just writing files to a regular filesystem on the NAS, it's not an object-storage